Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py

```diff
@@ -0,0 +1,159 @@
+    for any owner in the list
+
+    Expects `Count` key and replaces it with `Ranking Score`
+    """
+    if owners:
+        max_count = max(owner.get('Count', 1) for owner in owners)
+        for owner in owners:
+            count = owner.pop('Count', 1)
+            owner['Ranking Score'] = count / max_count
+    return owners
+
+
+def rank(owners: List[Dict[str, Any]], k: int = 5) -> List[Dict[str, Any]]:
+    """
+    Return up to k owners with the highest ranking scores
+    """
+    if k <= 0:
+        raise ValueError(f'Number of owners k={k} must be greater than zero')
+    return sorted(owners, key=lambda x: x['Ranking Score'])[:k]
+
+
+def justify(owners: List[Dict[str, str]]) -> List[Dict[str, str]]:
+    """
+    For now, `Justification` is the same as `Source`; in the future, will sophisticate
+    """
+    for owner in owners:
+        owner['Justification'] = owner.get('Source', '')
+    return owners
+
+
+def _canonicalize(owner: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Canonicalizes an owner dictionary and adds a deduplication key
+    `Canonicalization` whose value is either:
+        1. whitespace-stripped and lower-cased email, if email exists
+        2. whitespace-stripped and lower-cased name
+        3. empty string if neither exists
+    """
+    if owner.get('Email', ''):
+        owner['Canonicalization'] = owner['Email'].strip().lower()
+        owner['Email'] = owner['Canonicalization']
+    elif owner.get('Name', ''):
+        owner['Canonicalization'] = owner['Name'].strip().lower()
+        owner['Name'] = owner['Canonicalization']
+    else:
+        owner['Canonicalization'] = ''
+    return owner
+
+
+def canonicalize(owners: List[Dict[str, str]]) -> List[Dict[str, Any]]:
+    """
+    Calls _canonicalize on each owner.
+
+    Defensive handling of potentially missing or NoneType inputs
+    """
+    canonicalized = []
+    if owners:
+        for owner in owners:
+            if owner:
+                for key in ('Name', 'Source', 'Email', 'Timestamp'):
+                    if key not in owner or owner[key] is None:
+                        owner[key] = ''
+                canonicalized.append(_canonicalize(owner))
+    return canonicalized
+
+
+def aggregate(owners: List[Dict[str, str]]) -> List[Dict[str, Any]]:
+    """
+    Aggregate owners by their canonicalization.
+
+    If canonicalized form is email, preserve longest name.
+    Preserve max timestamp and union over sources.
+
+    Aggregate remaining keys by type: union over strings, and max over numerical types.
+    If type is neither of the above, all values of that key will be dropped from the aggregated owner.
+    """
+    deduped = []
+    sorted_owners = sorted(owners, key=lambda owner: owner['Canonicalization'])
+    for key, group in groupby(sorted_owners, key=lambda owner: owner['Canonicalization']):
+        duplicates = list(group)
+        if key == duplicates[0].get('Email', ''):
```

Comment:
is it possible to simplify with the observation that if the canonicalized form is name, we also are preserving the longest name? (the names might all be ["Bob", "Bob", "Bob"] but the single longest name is still valid)